cpp/src/traversal/mg/bfs.cuh

```diff
@@ -17,115 +17,119 @@
+         cugraph::GraphCSRView<vertex_t, edge_t, weight_t> const &graph,
+         vertex_t *distances,
+         vertex_t *predecessors,
+         const vertex_t start_vertex)
 {
-  using namespace detail;
+  CUGRAPH_EXPECTS(handle.comms_initialized(),
+                  "cugraph::mg::bfs() expected to work only in multi gpu case.");
 
-  // We need to keep track if a vertex is visited or its status
-  // This needs to be done for all the vertices in the global graph
   size_t word_count = detail::number_of_words(graph.number_of_vertices);
-  rmm::device_vector<unsigned> input_frontier(word_count);
-  rmm::device_vector<unsigned> output_frontier(word_count);
-  rmm::device_vector<unsigned> visited(word_count);
+  rmm::device_vector<unsigned> isolated_bmap(word_count, 0);
+  rmm::device_vector<unsigned> visited_bmap(word_count, 0);
+  rmm::device_vector<unsigned> output_frontier_bmap(word_count, 0);
 
-  rmm::device_vector<unsigned> frontier_not_empty(1);
+  rmm::device_vector<unsigned> unique_bmap(word_count, 0);
 
-  // Load balancer for calls to bfs functors
-  LoadBalanceExecution<VT, ET, WT> lb(handle, graph);
+  // Buffers required for BFS
+  rmm::device_vector<vertex_t> input_frontier(graph.number_of_vertices);
+  rmm::device_vector<vertex_t> output_frontier(graph.number_of_vertices);
+  rmm::device_vector<size_t> temp_buffer_len(handle.get_comms().get_size());
 
-  // Functor to check if frontier is empty
-  is_not_equal neq(static_cast<unsigned>(0), frontier_not_empty.data().get());
+  // Frontier Expand for calls to bfs functors
+  detail::FrontierExpand<vertex_t, edge_t, weight_t> fexp(handle, graph);
 
   cudaStream_t stream = handle.get_stream();
 
-  // Fill predecessors with an invalid vertex id
+  // Reusing buffers to create isolated bitmap
+  {
+    rmm::device_vector<vertex_t> &local_isolated_ids  = input_frontier;
+    rmm::device_vector<vertex_t> &global_isolated_ids = output_frontier;
+    detail::create_isolated_bitmap(
+      handle, graph, local_isolated_ids, global_isolated_ids, temp_buffer_len, isolated_bmap);
+  }
+
+  // Initialize input frontier
+  input_frontier[0]           = start_vertex;
+  vertex_t input_frontier_len = 1;
+
+  vertex_t level = 0;
+  if (distances != nullptr) { detail::fill_max_dist(handle, graph, start_vertex, distances); }
   thrust::fill(rmm::exec_policy(stream)->on(stream),
                predecessors,
                predecessors + graph.number_of_vertices,
-               graph.number_of_vertices);
+               cugraph::invalid_idx<vertex_t>::value);
 
-  VT level = 0;
-  if (distances != nullptr) {
-    thrust::fill(rmm::exec_policy(stream)->on(stream),
-                 distances,
-                 distances + graph.number_of_vertices,
-                 std::numeric_limits<VT>::max());
-  }
-
-  // BFS communications wrapper
-  BFSCommunicatorBCastReduce<VT, ET, WT> bfs_comm(handle, word_count);
+  // TODO : Check if starting vertex is isolated. Exit function if it is.
```

Comment:
TODO here. It should be addressed or turned into a FIXME so that it can be captured in tech debt.

Reply:
I will address it in this PR